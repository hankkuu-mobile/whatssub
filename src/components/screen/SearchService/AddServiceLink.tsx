import * as React from 'react';

import styled from 'styled-components/native';

export interface AddServiceLinkProps {
  onClickAddServiceLink: () => void;
  style?: any;
}

const Container = styled.View`
  background-color: ${({ theme }) => (theme.debug ? 'salmon' : 'transparent')};
  height: 50;
  justify-content: center;
  align-items: center;
`;

const Touchable = styled.TouchableOpacity`
  background-color: ${({ theme }) => (theme.debug ? '#40E0D0' : 'transparent')};
  padding-top: 7;
  padding-bottom: 7;
  padding-left: 10;
  padding-right: 10;
`;

const StyledText = styled.Text`
  color: #407fff;
`;

function AddServiceLink({
  onClickAddServiceLink,
  ...rest
}: AddServiceLinkProps) {
  return (
    <Container {...rest}>
      <Touchable
        testID={'TOUCHABLE_ADD_SERVICE_LINK'}
        onPress={onClickAddServiceLink}
      >
        <StyledText>
          {'+ 검색 결과가 없다면 여기를 눌러 직접 입력해보세요.'}
        </StyledText>
      </Touchable>
    </Container>
  );
}

export default AddServiceLink;
