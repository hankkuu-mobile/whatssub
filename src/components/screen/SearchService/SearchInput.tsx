import * as React from 'react';

import { IC_MAGNIFIER } from '../../../utils/Icons';
import styled from 'styled-components/native';

const Container = styled.View`
  background-color: ${({ theme }) => (theme.debug ? '#ddd' : '#f3f5f7')};
  height: 42;
  margin-left: 20;
  margin-right: 20;
  border-radius: 8;

  flex-direction: row;
`;

const MagContainer = styled.View`
  background-color: ${({ theme }) => (theme.debug ? 'red' : 'transparent')};
  height: 24;
  width: 24;
  margin-left: 8;
  margin-right: 2;
  margin-top: 9;
  margin-bottom: 9;
  justify-content: center;
  align-items: center;
`;

const Magnifier = styled.Image`
  background-color: ${({ theme }) => (theme.debug ? 'green' : 'transparent')};
  width: 16;
  height: 16;
  margin-bottom: 2;
`;

const Input = styled.TextInput`
  background-color: ${({ theme }) => (theme.debug ? '#aaa' : 'transparent')};
  flex-grow: 1;
  align-self: center;
  font-size: 16;
`;

const ResetContainer = styled.View`
  background-color: ${({ theme }) => (theme.debug ? 'blue' : 'transparent')};
  height: 24;
  width: 24;
  margin-left: 2;
  margin-right: 11;
  margin-top: 9;
  margin-bottom: 9;
  justify-content: center;
  align-items: center;
`;

const Reset = styled.TouchableOpacity`
  background-color: ${({ theme }) => (theme.debug ? 'yellow' : '#c6ccd1')};
  border-radius: 20;
  width: 20;
  height: 20;
  justify-content: center;
  align-items: center;
`;

const ResetText = styled.Text`
  color: white;
`;

interface SearchInputProps {
  value: string;
  onDebounceOrOnReset: (value: string) => void;
  style?: any;
}

// reference : https://dev.to/gabe_ragland/debouncing-with-react-hooks-jci
function useDebounce(value, delay) {
  const [debouncedValue, setDebouncedValue] = React.useState(value);

  React.useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);
    return () => {
      clearTimeout(handler);
    };
  }, [value]);

  return debouncedValue;
}

function SearchInput(props: SearchInputProps) {
  const [value, setValue] = React.useState(props.value);
  const debouncedValue = useDebounce(value, 400);

  React.useEffect(() => {
    props.onDebounceOrOnReset(debouncedValue);
  }, [debouncedValue]);

  React.useEffect(() => {
    setValue(props.value);
  }, [props.value]);

  return (
    <Container style={props.style}>
      <MagContainer>
        <Magnifier source={IC_MAGNIFIER} />
      </MagContainer>
      {/* TODO: add placeholder string to Sting */}
      <Input
        testID={'SEARCH_INPUT'}
        value={value}
        onChangeText={(text) => {
          setValue(text);
        }}
        placeholder={'검색어를 입력해주세요.'}
        placeholderTextColor={'#cdd2d7'}
      />
      {props.value !== '' && (
        <ResetContainer>
          <Reset
            testID='RESET_BUTTON'
            onPress={() => {
              props.onDebounceOrOnReset('');
            }}
          >
            <ResetText>X</ResetText>
          </Reset>
        </ResetContainer>
      )}
    </Container>
  );
}

export default SearchInput;
