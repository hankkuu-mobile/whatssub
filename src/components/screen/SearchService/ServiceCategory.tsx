import * as React from 'react';

import { getString } from '../../../../STRINGS';
import styled from 'styled-components/native';

const Container = styled.View`
  background-color: ${({ theme }) => (theme.debug ? '#ddd' : 'transparent')};
  margin-left: 20;
  margin-right: 20;
  flex-direction: row;
  justify-content: flex-start;
  flex-wrap: wrap;
`;

const HorizontalScrollView = styled.ScrollView`
  padding-top: 12;
  padding-bottom: 12;
`;

const Category = styled.TouchableOpacity<{ selected?: boolean }>`
  background-color: ${({ theme, selected }) =>
    theme.debug ? 'red' : selected ? '#3e7eff' : 'transparent'};
  border-width: 1;
  border-color: #e1e4e7;
  border-radius: 17;
  padding-top: 9;
  padding-bottom: 9;
  padding-left: 15;
  padding-right: 15;
  margin-right: 5;
  justify-content: center;
  align-items: center;
`;

const Label = styled.Text<{ selected?: boolean }>`
  color: ${({ selected }) => (selected ? 'white' : '#a4a5a6')};
  font-size: 12;
  letter-spacing: -0.5;
`;

export interface SearchInputProps {
  categoryList: string[];
  initialCategory?: string;
  onSelect?: (category: string) => void;
  style?: any;
}

// TODO: error handle when there is no props.categoryList exists
function ServiceCategory(props: SearchInputProps) {
  const [selected, select] = React.useState(
    props.initialCategory ? props.initialCategory : '',
  );

  // TODO: Add assertion that category is unique. as it is being used as key
  return (
    <Container style={props.style}>
      <HorizontalScrollView horizontal={true}>
        {props.categoryList.map((category) => (
          <Category
            testID={`CATEGORY_SELECT_${category}`}
            key={category}
            selected={category === selected}
            onPress={() => {
              select(category);
              props.onSelect && props.onSelect(category);
            }}
          >
            <Label selected={category === selected}>
              {getString(category)}
            </Label>
          </Category>
        ))}
      </HorizontalScrollView>
    </Container>
  );
}

export default ServiceCategory;
