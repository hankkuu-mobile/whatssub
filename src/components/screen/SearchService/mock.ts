const THUMBNAIL =
  'https://facebook.github.io/react-native/docs/assets/favicon.png';

export interface Service {
  id: string;
  name: string;
  category: string;
  thumbnail: string;
  initialGroup: string;
}

// category key is already localized
export const SERVICE_LIST: Service[] = [
  {
    id: '1',
    name: '휴대폰ㆍ인터넷ㆍTV',
    category: 'CATEGORY_COMMUNICATION_EXPENSE',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '2',
    name: '대출ㆍ보험ㆍ적금ㆍ예금ㆍ곗돈',
    category: 'CATEGORY_FINANCE',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '3',
    name: '공과금ㆍ렌탈',
    category: 'CATEGORY_RESIDENCE_AND_MAINTENANCE',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '4',
    name: '학원ㆍ기부ㆍ헌금',
    category: 'CATEGORY_ETC',
    thumbnail: THUMBNAIL,
    initialGroup: '0',
  },
  {
    id: '5',
    name: '넷플릭스',
    category: 'CATEGORY_ENTERTAINMENT',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '6',
    name: '에버노트',
    category: 'CATEGORY_PRODUCTIVITY',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '7',
    name: '폴라리스 오피스',
    category: 'CATEGORY_PRODUCTIVITY',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '8',
    name: '유투브 프리미엄',
    category: 'CATEGORY_ENTERTAINMENT',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
  {
    id: '9',
    name: '멜론',
    category: 'CATEGORY_MUSIC',
    thumbnail: THUMBNAIL,
    initialGroup: '1',
  },
];

export const EXCEPTIONAL_CATEGORY_CASE = 'CATEGORY_RECOMMENDED';
export const CATEGORY_LIST = [
  'CATEGORY_COMMUNICATION_EXPENSE',
  'CATEGORY_FINANCE',
  'CATEGORY_RESIDENCE_AND_MAINTENANCE',
  'CATEGORY_ETC',
  'CATEGORY_ENTERTAINMENT',
  'CATEGORY_PRODUCTIVITY',
  'CATEGORY_MUSIC',
];
// push the exceptional case to the first index
CATEGORY_LIST.unshift(EXCEPTIONAL_CATEGORY_CASE);

export interface SeviceGroupTable {
  [key: string]: { key: string; order: number };
}
// title order lookup table
// `order` should starts from 0 and consecutively
// `order` is being used when initial case
export const INITIAL_GROUP_ORDER_TABLE: SeviceGroupTable = {
  0: { key: 'SERVICE_GROUP_LIVING_EXPENSE', order: 0 },
  1: { key: 'SERVICE_GROUP_SUBSCRIPTION_SERVICE', order: 1 },
};
