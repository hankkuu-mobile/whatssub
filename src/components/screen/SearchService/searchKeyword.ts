export function searchKeyword<T extends object>(
  obj: T,
  keyword: string,
  keys: Array<keyof T>,
  caseSensitive?: boolean,
) {
  for (const [curr, currElem] of Object.entries(obj)) {
    if (keys.find((el) => el === curr)) {
      if (
        (caseSensitive ? currElem : currElem.toLowerCase()).includes(
          caseSensitive ? keyword : keyword.toLowerCase(),
        )
      ) {
        return true;
      }
    }
  }
  return false;
}
