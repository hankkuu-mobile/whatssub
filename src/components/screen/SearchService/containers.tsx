import { CATEGORY_LIST, INITIAL_GROUP_ORDER_TABLE, SERVICE_LIST } from './mock';

import { withContainer } from '../../../utils/hocs';

// used at Test
export const MOCK = {
  serviceList: SERVICE_LIST,
  initialServiceGroupTable: INITIAL_GROUP_ORDER_TABLE,
  categoryList: CATEGORY_LIST,
};

// used at storybook for now
// we might be able to use this(withContainer()) as a pattern for injecting real data later
const withSearchServiceMock = withContainer({
  ...MOCK,
});

export { withSearchServiceMock };
