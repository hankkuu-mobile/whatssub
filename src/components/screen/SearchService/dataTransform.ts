import { Service, SeviceGroupTable } from './mock';

import { getString } from '../../../../STRINGS';

export interface ServiceGroup {
  title: string;
  data: { id: string; name: string; category: string; thumbnail: string }[];
  order: number;
}

export const dataTransform = (
  serviceList: Service[],
  serviceGroupTable: SeviceGroupTable,
) => {
  const serviceGroup = serviceList.reduce((acc: ServiceGroup[], service) => {
    const group = serviceGroupTable[service.initialGroup];
    const isServiceGroupExist = acc.find((obj) => obj.title === group.key);

    if (isServiceGroupExist) {
      isServiceGroupExist.data.push(service);
    } else {
      acc.push({
        title: group.key,
        data: [service],
        order: group.order,
      });
    }
    return acc;
  }, []);

  return serviceGroup.sort((a, b) => a.order - b.order);
};
