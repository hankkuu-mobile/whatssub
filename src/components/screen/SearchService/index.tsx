import {
  CATEGORY_LIST,
  EXCEPTIONAL_CATEGORY_CASE,
  INITIAL_GROUP_ORDER_TABLE,
  SERVICE_LIST,
  Service,
  SeviceGroupTable,
} from './mock';
import React, { useEffect, useState } from 'react';

import AddServiceLink from './AddServiceLink';
import { NavigationScreenProp } from 'react-navigation';
import SearchInput from './SearchInput';
import ServiceCategory from './ServiceCategory';
import ServiceList from './ServiceList';
import { dataTransform } from './dataTransform';
import { getString } from '../../../../STRINGS';
import { searchKeyword } from './searchKeyword';
import styled from 'styled-components/native';
import { withSearchServiceMock } from './containers';

/**
 * Get service list from server and process it with categorizing
 * and input keyword filtering. then transform its data as
 * the form of sectionList to show on UI
 */

const Container = styled.View`
  flex: 1;
`;

export interface Props {
  navigation?: NavigationScreenProp<any, any>;
  initialInputValue?: string;
  serviceList: Service[];
  initialServiceGroupTable: SeviceGroupTable;
  categoryList: string[];
  // navigate to Add Service Screen
  onClickAddService: () => void;
}

const localizeCategoryProperty = (obj: Service): Service =>
  Object.assign<{}, Service, Pick<Service, 'category'>>({}, obj, {
    category: getString(obj.category),
  });

function SearchService(props: Props) {
  const [inputValue, setInputValue] = useState(
    props.initialInputValue ? props.initialInputValue : '',
  );
  const [category, setCategory] = useState(EXCEPTIONAL_CATEGORY_CASE);
  const [serviceListOfSectionList, setServiceListOfSectionList] = useState([]);

  useEffect(() => {
    let filteredServiceList;
    if (inputValue !== '' || category === EXCEPTIONAL_CATEGORY_CASE) {
      filteredServiceList = props.serviceList.filter((service) => {
        return searchKeyword<Service>(
          localizeCategoryProperty(service),
          inputValue,
          ['name', 'category'],
        );
      });
    } else {
      filteredServiceList = props.serviceList.filter((service) => {
        return searchKeyword<Service>(
          localizeCategoryProperty(service),
          getString(category),
          ['category'],
        );
      });
    }

    setServiceListOfSectionList(
      dataTransform(filteredServiceList, props.initialServiceGroupTable),
    );
  }, [category, inputValue]);

  return (
    <Container>
      <SearchInput value={inputValue} onDebounceOrOnReset={setInputValue} />
      <ServiceCategory
        style={inputValue !== '' && { display: 'none' }}
        categoryList={props.categoryList}
        initialCategory={category}
        onSelect={(category) => setCategory(category)}
      />
      {inputValue !== '' && (
        <AddServiceLink onClickAddServiceLink={props.onClickAddService} />
      )}
      <ServiceList serviceList={serviceListOfSectionList} />
    </Container>
  );
}

export default SearchService;

const SearchServiceWithMock = withSearchServiceMock(SearchService);

export { SearchServiceWithMock };
