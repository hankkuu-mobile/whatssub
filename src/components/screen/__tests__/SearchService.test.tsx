import 'react-native';

import * as React from 'react';

import { Props, SearchServiceWithMock } from '../SearchService';
import {
  fireEvent,
  getAllByTestId,
  render,
  wait,
  waitForElement,
} from '@testing-library/react-native';
import renderer, { act, create } from 'react-test-renderer';

import { EXCEPTIONAL_CATEGORY_CASE } from '../SearchService/mock';
import { InitialProviders } from '../../../providers';
import { MOCK } from '../SearchService/containers';
import { getString } from '../../../../STRINGS';
import i18n from 'i18n-js';

i18n.locale = 'ko';

const jestFn = {
  onClickAddService: jest.fn(),
};

const createTestProps = (props: Partial<Props>): Props => ({
  ...MOCK,
  ...jestFn, // override any function from MOCK with jest.fn()
  initialInputValue: '', // custom initial props
  ...props, // override anything as user props
});

const testComponent = (props: Props) => (
  <InitialProviders doNotWaitFont>
    <SearchServiceWithMock {...props} />
  </InitialProviders>
);

describe('[SearchService] render', () => {
  it('component and snapshot matches', () => {
    const rendered: renderer.ReactTestRendererJSON = renderer
      .create(testComponent(createTestProps({})))
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[SearchService] interaction', () => {
  it('should contain service element when searching by its name', () => {
    const toFilter = MOCK.serviceList[0];
    const rendered = render(
      testComponent(createTestProps({ initialInputValue: toFilter.name })),
    );
    const names = rendered.getAllByTestId('SERVICE_ELEMENT_NAME');
    names.map((name) => expect(name.props.children).toContain(toFilter.name));
  });

  it('should contain relevant service element when category is clicked', async () => {
    const { getByTestId, getAllByTestId } = render(
      testComponent(createTestProps({})),
    );

    // *the index can be changed manually by user's needs
    const indexToTest = 0;

    // remove exceptional case from category list
    const categoriesWithoutExceptionalCase = MOCK.categoryList.filter(
      (cat) => cat !== EXCEPTIONAL_CATEGORY_CASE,
    );

    const category = categoriesWithoutExceptionalCase[indexToTest];
    // get category dom
    const categoryDom = getByTestId(`CATEGORY_SELECT_${category}`);
    // click category dom
    fireEvent.press(categoryDom);
    // wait until service element list has relevant element
    const categoryListOfFilteredServiceElement = await waitForElement(() => {
      const categoryListOfServiceElement = getAllByTestId(
        'SERVICE_ELEMENT_CATEGORY',
      );
      return categoryListOfServiceElement.filter((se) => {
        if (se.props.children !== getString(category)) {
          throw Error(`unexpected service elements are included`);
        } else {
          return true;
        }
      });
    });

    // check if the categories of filtered service element are desired category
    categoryListOfFilteredServiceElement.map((se) => {
      expect(se.props.children).toBe(getString(category));
    });

    // Todo: for better test in the future, you might need to check
    // whether all `category` are actually exists in SERVICE_LIST or not
  });
});
