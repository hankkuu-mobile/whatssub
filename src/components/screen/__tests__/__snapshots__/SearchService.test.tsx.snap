// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SearchService] render component and snapshot matches 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#f3f5f7",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 42,
          "marginLeft": 20,
          "marginRight": 20,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "height": 24,
            "justifyContent": "center",
            "marginBottom": 9,
            "marginLeft": 8,
            "marginRight": 2,
            "marginTop": 9,
            "width": 24,
          },
        ]
      }
    >
      <Image
        source={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "height": 16,
              "marginBottom": 2,
              "width": 16,
            },
          ]
        }
      />
    </View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="검색어를 입력해주세요."
      placeholderTextColor="#cdd2d7"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "transparent",
            "flexGrow": 1,
            "fontSize": 16,
          },
        ]
      }
      testID="SEARCH_INPUT"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
          "marginLeft": 20,
          "marginRight": 20,
        },
        false,
      ]
    }
  >
    <RCTScrollView
      horizontal={true}
      style={
        Array [
          Object {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
        ]
      }
    >
      <View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#3e7eff",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_RECOMMENDED"
        >
          <Text
            selected={true}
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            추천
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_COMMUNICATION_EXPENSE"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            통신비
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_FINANCE"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            금융
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_RESIDENCE_AND_MAINTENANCE"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            주거/관리
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_ETC"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            기타
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_ENTERTAINMENT"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            엔터테인먼트
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_PRODUCTIVITY"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            생산성
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#e1e4e7",
              "borderRadius": 17,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginRight": 5,
              "opacity": 1,
              "paddingBottom": 9,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 9,
            }
          }
          testID="CATEGORY_SELECT_CATEGORY_MUSIC"
        >
          <Text
            selected={false}
            style={
              Array [
                Object {
                  "color": "#a4a5a6",
                  "fontSize": 12,
                  "letterSpacing": -0.5,
                },
              ]
            }
          >
            음악
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <RCTScrollView
      ListFooterComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "displayName": "Styled(View)",
          "inlineStyle": InlineStyle {
            "rules": Array [
              "
  display: flex;
  width: 100%;
  height: 1px;
  background-color: rgb(230, 234, 238);
",
            ],
          },
          "render": [Function],
          "styledComponentId": "StyledNativeComponent",
          "target": [Function],
          "withComponent": [Function],
        }
      }
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      renderSectionHeader={[Function]}
      scrollEventThrottle={50}
      sections={Array []}
      stickyHeaderIndices={Array []}
      stickySectionHeadersEnabled={true}
      style={
        Array [
          Object {
            "backgroundColor": "rgb(249, 250, 251)",
          },
        ]
      }
      updateCellsBatchingPeriod={50}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "rgb(230, 234, 238)",
                  "display": "flex",
                  "height": 1,
                  "width": "100%",
                },
              ]
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
